

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>plotCSprofile &mdash; Try Sphinx V1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../_static/custom.css?v=c4e650ba" />

  
      <script src="../../_static/documentation_options.js?v=d43430bd"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            Try Sphinx
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../list_of_tools.html">The tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../example_usage.html">Usage Manual</a></li>
<li class="toctree-l1"><a class="reference internal" href="../about.html">About</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Try Sphinx</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">plotCSprofile</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/content/tools/plotCSprofile.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="plotcsprofile">
<h1>plotCSprofile<a class="headerlink" href="#plotcsprofile" title="Link to this heading">¶</a></h1>
<p>This tool calculates scores per genome regions and creates a profile plot for scores over sets of genomic regions.
Typically, the genome regions are genes, but any other regions defined in a BED file can be used. A chromatin state
segmentation generated by chromIDEAS or ideasCS is required.</p>
<p>To learn more about the specific parameters, type:</p>
<div class="highlight-sh notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>plotCSprofile<span class="w"> </span>TSS<span class="w"> </span>-h
OR
$<span class="w"> </span>plotCSprofile<span class="w"> </span>Body<span class="w"> </span>-h
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>plotCSprofile TSS: In TSS mode, only those genomicpositions before (upstream) and/or after (downstream) of the TSS will be plotted.</p></li>
<li><p>plotCSprofile Body: In Body mode, all regions are stretched or shrunken to the length indicated by the user.</p></li>
</ul>
</div>
</section>
<section id="content">
<h1>Content<a class="headerlink" href="#content" title="Link to this heading">¶</a></h1>
<nav class="contents local" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#plotcsprofile-tss-mode" id="id3">plotCSprofile TSS Mode</a></p>
<ul>
<li><p><a class="reference internal" href="#required-arguments" id="id4">Required arguments</a></p></li>
<li><p><a class="reference internal" href="#optional-arguments" id="id5">Optional arguments</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#plotcsprofile-body-mode" id="id6">plotCSprofile Body Mode</a></p>
<ul>
<li><p><a class="reference internal" href="#id1" id="id7">Required arguments</a></p></li>
<li><p><a class="reference internal" href="#id2" id="id8">Optional arguments</a></p></li>
</ul>
</li>
</ul>
</nav>
<section id="plotcsprofile-tss-mode">
<h2><a class="toc-backref" href="#id3" role="doc-backlink">plotCSprofile TSS Mode</a><a class="headerlink" href="#plotcsprofile-tss-mode" title="Link to this heading">¶</a></h2>
<section id="required-arguments">
<h3><a class="toc-backref" href="#id4" role="doc-backlink">Required arguments</a><a class="headerlink" href="#required-arguments" title="Link to this heading">¶</a></h3>
<dl>
<dt><code class="docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;input&gt;</span></code></dt><dd><p>chromatin state segmentation file (space delimited) generated by chromIDEAS or ideasCS <strong>[Default: None]</strong>. Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#ID CHR POSst POSed cell1 cell2</span>
<span class="mi">1</span> <span class="n">chr1</span> <span class="mi">792600</span> <span class="mi">792800</span> <span class="mi">1</span> <span class="mi">1</span>
<span class="mi">2</span> <span class="n">chr1</span> <span class="mi">792800</span> <span class="mi">793000</span> <span class="mi">0</span> <span class="mi">0</span>
<span class="mi">3</span> <span class="n">chr1</span> <span class="mi">793000</span> <span class="mi">793200</span> <span class="mi">0</span> <span class="mi">0</span>
<span class="mi">4</span> <span class="n">chr1</span> <span class="mi">793200</span> <span class="mi">793400</span> <span class="mi">0</span> <span class="mi">0</span>
</pre></div>
</div>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;output_figure&gt;</span></code></dt><dd><p>File name to save the image. The file ending will be used to determine the image format. The available options are: “png”, “jpg”, “pdf” and “svg”, e.g., MyHeatmap.png. <strong>[Default: None]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-r</span> <span class="pre">&lt;regionFile&gt;</span></code></dt><dd><p>File name or names, in BED or GTF format, containing the regions to plot. The GTF format is recommended. To ensure result accuracy, for BED format inputs, it is required to include strand information in column 4. If the strand is uncertain, use ‘*’ to represent it <strong>[Default: None]</strong>. Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chr1</span>  <span class="mi">100</span>  <span class="mi">500</span>  <span class="o">+</span>
<span class="n">chr3</span>  <span class="mi">600</span>  <span class="mi">900</span>  <span class="o">*</span>
<span class="o">...</span>
</pre></div>
</div>
</dd>
</dl>
</section>
<section id="optional-arguments">
<h3><a class="toc-backref" href="#id5" role="doc-backlink">Optional arguments</a><a class="headerlink" href="#optional-arguments" title="Link to this heading">¶</a></h3>
<dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">-S</span> <span class="pre">&lt;sampleID&gt;</span></code></dt><dd><p>Specify the sample ID(s) to plot the distribution curve. The ID(s) must match those in <code class="docutils literal notranslate"><span class="pre">&lt;input&gt;</span></code>. If multiple sample IDs need to be specified simultaneously, separate them with commas, e.g., “sample1,sample2,sample3”. <strong>[Default: All]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">&lt;output_matrix&gt;</span></code></dt><dd><p>File name to save the underlying data for the chromatin state profile. Users can use this matrix to replot the data with other software (e.g., R). <strong>[Default: None]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-t</span> <span class="pre">&lt;region_type&gt;</span></code></dt><dd><p>When <code class="docutils literal notranslate"><span class="pre">&lt;regionFile&gt;</span></code> is specified as a GTF file, select whether to use gene or transcript coordinates for analysis. <strong>[Default: transcript]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-u</span> <span class="pre">&lt;up_num&gt;</span></code></dt><dd><p>Number of bins covering the upstream region of the transcription start site. <strong>[Default: 5]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-d</span> <span class="pre">&lt;down_num&gt;</span></code></dt><dd><p>Number of bins covering the downstream region of the transcription end site. <strong>[Default: 5]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-l</span> <span class="pre">&lt;refPointLabel&gt;</span></code></dt><dd><p>only for TSS mode] Label shown in the plot for the TSS. Default is the <strong>TSS</strong>, but could be anything, e.g. “peak start”.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-z</span> <span class="pre">&lt;scale_type&gt;</span></code></dt><dd><p>Mode for calculating chromatin state proportions. Possible choices: “genomic” and “state”. “genomic” will compute the local proportion of each chromatin state at every site. (Example: For Site 1 with 100 “S1” out of 1,000 total values → “S1” = 100/1,000 = 10%); “state” will compute the global contribution of each site to a state’s total occurrences. (Example: If “S1” occurs 50,000 times genome-wide, Site 1’s 100 “S1” values contribute :100/50,000 = 0.2%). <strong>[Default: genomic]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-W</span> <span class="pre">&lt;width&gt;</span></code></dt><dd><p>Plot width in Inch. <strong>[Default: 10]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-H</span> <span class="pre">&lt;height&gt;</span></code></dt><dd><p>Plot height in Inch. <strong>[Default: 8]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-C</span> <span class="pre">&lt;colors&gt;</span></code></dt><dd><p>List of colors to use for the plotted lines. Color names and html hex strings (e.g., #eeff22) are accepted. The color names should be commas separated, e.g., red,blue. <strong>[Default: Auto]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-N</span> <span class="pre">&lt;num_per_row&gt;</span></code></dt><dd><p>Number of plots per row. <strong>[Default: 5]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-O</span> <span class="pre">&lt;order_of_CS&gt;</span></code></dt><dd><p>Plot ordering scheme. Available options: (1) By chromatin state name (cs); (2) By chromatin state genomic percentage (csgp); (3) Custom order (comma-separated, e.g., s3,s1,s2). <strong>[Default: cs]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-R</span></code></dt><dd><p>Reverse the plot ordering scheme.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-p</span> <span class="pre">&lt;nthreads&gt;</span></code></dt><dd><p>Number of parallel processes. <strong>[Default: 4]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-h</span></code></dt><dd><p>Show this help message and exit.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-v</span></code></dt><dd><p>Show program’s version number and exit.</p>
</dd>
</dl>
</section>
</section>
<section id="plotcsprofile-body-mode">
<h2><a class="toc-backref" href="#id6" role="doc-backlink">plotCSprofile Body Mode</a><a class="headerlink" href="#plotcsprofile-body-mode" title="Link to this heading">¶</a></h2>
<section id="id1">
<h3><a class="toc-backref" href="#id7" role="doc-backlink">Required arguments</a><a class="headerlink" href="#id1" title="Link to this heading">¶</a></h3>
<dl>
<dt><code class="docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;input&gt;</span></code></dt><dd><p>chromatin state segmentation file (space delimited) generated by chromIDEAS or ideasCS <strong>[Default: None]</strong>. Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#ID CHR POSst POSed cell1 cell2</span>
<span class="mi">1</span> <span class="n">chr1</span> <span class="mi">792600</span> <span class="mi">792800</span> <span class="mi">1</span> <span class="mi">1</span>
<span class="mi">2</span> <span class="n">chr1</span> <span class="mi">792800</span> <span class="mi">793000</span> <span class="mi">0</span> <span class="mi">0</span>
<span class="mi">3</span> <span class="n">chr1</span> <span class="mi">793000</span> <span class="mi">793200</span> <span class="mi">0</span> <span class="mi">0</span>
<span class="mi">4</span> <span class="n">chr1</span> <span class="mi">793200</span> <span class="mi">793400</span> <span class="mi">0</span> <span class="mi">0</span>
</pre></div>
</div>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;output_figure&gt;</span></code></dt><dd><p>File name to save the image. The file ending will be used to determine the image format. The available options are: “png”, “jpg”, “pdf” and “svg”, e.g., MyHeatmap.png. <strong>[Default: None]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-r</span> <span class="pre">&lt;regionFile&gt;</span></code></dt><dd><p>File name or names, in BED or GTF format, containing the regions to plot. The GTF format is recommended. To ensure result accuracy, for BED format inputs, it is required to include strand information in column 4. If the strand is uncertain, use ‘*’ to represent it <strong>[Default: None]</strong>. Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chr1</span>  <span class="mi">100</span>  <span class="mi">500</span>  <span class="o">+</span>
<span class="n">chr3</span>  <span class="mi">600</span>  <span class="mi">900</span>  <span class="o">*</span>
<span class="o">...</span>
</pre></div>
</div>
</dd>
</dl>
</section>
<section id="id2">
<h3><a class="toc-backref" href="#id8" role="doc-backlink">Optional arguments</a><a class="headerlink" href="#id2" title="Link to this heading">¶</a></h3>
<dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">-S</span> <span class="pre">&lt;sampleID&gt;</span></code></dt><dd><p>Specify the sample ID(s) to plot the distribution curve. The ID(s) must match those in <code class="docutils literal notranslate"><span class="pre">&lt;input&gt;</span></code>. If multiple sample IDs need to be specified simultaneously, separate them with commas, e.g., “sample1,sample2,sample3”. <strong>[Default: All]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">&lt;output_matrix&gt;</span></code></dt><dd><p>File name to save the underlying data for the chromatin state profile. Users can use this matrix to replot the data with other software (e.g., R). <strong>[Default: None]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-t</span> <span class="pre">&lt;region_type&gt;</span></code></dt><dd><p>When <code class="docutils literal notranslate"><span class="pre">&lt;regionFile&gt;</span></code> is specified as a GTF file, select whether to use gene or transcript coordinates for analysis. <strong>[Default: transcript]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-b</span> <span class="pre">&lt;body_num&gt;</span></code></dt><dd><p>only for Body mode] Specify the number of bins for dividing the gene body region. <strong>[Default: 10]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-u</span> <span class="pre">&lt;up_num&gt;</span></code></dt><dd><p>Number of bins covering the upstream region of the transcription start site. <strong>[Default: 5]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-d</span> <span class="pre">&lt;down_num&gt;</span></code></dt><dd><p>Number of bins covering the downstream region of the transcription end site. <strong>[Default: 5]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-s</span> <span class="pre">&lt;startLabel&gt;</span></code></dt><dd><p>only for Body mode] Label shown in the plot for the start of the region. <strong>[Default: TSS]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-e</span> <span class="pre">&lt;endLabel&gt;</span></code></dt><dd><p>only for Body mode] Label shown in the plot for the region end. <strong>[Default: TES]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-z</span> <span class="pre">&lt;scale_type&gt;</span></code></dt><dd><p>Mode for calculating chromatin state proportions. Possible choices: “genomic” and “state”. “genomic” will compute the local proportion of each chromatin state at every site. (Example: For Site 1 with 100 “S1” out of 1,000 total values → “S1” = 100/1,000 = 10%); “state” will compute the global contribution of each site to a state’s total occurrences. (Example: If “S1” occurs 50,000 times genome-wide, Site 1’s 100 “S1” values contribute :100/50,000 = 0.2%). <strong>[Default: genomic]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-W</span> <span class="pre">&lt;width&gt;</span></code></dt><dd><p>Plot width in Inch. <strong>[Default: 10]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-H</span> <span class="pre">&lt;height&gt;</span></code></dt><dd><p>Plot height in Inch. <strong>[Default: 8]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-C</span> <span class="pre">&lt;colors&gt;</span></code></dt><dd><p>List of colors to use for the plotted lines. Color names and html hex strings (e.g., #eeff22) are accepted. The color names should be commas separated, e.g., red,blue. <strong>[Default: Auto]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-N</span> <span class="pre">&lt;num_per_row&gt;</span></code></dt><dd><p>Number of plots per row. <strong>[Default: 5]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-O</span> <span class="pre">&lt;order_of_CS&gt;</span></code></dt><dd><p>Plot ordering scheme. Available options: (1) By chromatin state name (cs); (2) By chromatin state genomic percentage (csgp); (3) Custom order (comma-separated, e.g., s3,s1,s2). <strong>[Default: cs]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-R</span></code></dt><dd><p>Reverse the plot ordering scheme.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-p</span> <span class="pre">&lt;nthreads&gt;</span></code></dt><dd><p>Number of parallel processes. <strong>[Default: 4]</strong></p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-h</span></code></dt><dd><p>Show this help message and exit.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">-v</span></code></dt><dd><p>Show program’s version number and exit.</p>
</dd>
</dl>
</section>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, fatyang.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>